libs/wallet-ui/src/components/interactions/provider.tsx

```diff
@@ -0,0 +1,288 @@
+      q.set(event.traceID, {
+        traceID: event.traceID,
+        workflow: 'PERMISSION_REQUEST',
+        hostname: event.data.hostname,
+        permissions: event.data.permissions,
+        wallet: event.data.wallet,
+      })
+      break
+    }
+    // Transactions
+    case 'REQUEST_TRANSACTION_REVIEW_FOR_SENDING': {
+      const transaction = parseTransactionInput(event)
+
+      q.set(event.traceID, {
+        traceID: event.traceID,
+        workflow: 'TRANSACTION_REVIEW',
+        transaction,
+      })
+      dispatch({
+        type: 'ADD_TRANSACTION',
+        transaction,
+      })
+
+      break
+    }
+    case 'TRANSACTION_FAILED': {
+      if (
+        queueItem?.workflow === 'TRANSACTION_REVIEW' &&
+        queueItem.transaction
+      ) {
+        const transaction = parseTransactionInput(event, queueItem.transaction)
+
+        q.set(event.traceID, {
+          traceID: event.traceID,
+          workflow: 'TRANSACTION_REVIEW',
+          transaction,
+        })
+        dispatch({
+          type: 'UPDATE_TRANSACTION',
+          transaction,
+        })
+      }
+
+      break
+    }
+    case 'TRANSACTION_SUCCEEDED': {
+      if (
+        queueItem?.workflow === 'TRANSACTION_REVIEW' &&
+        queueItem.transaction
+      ) {
+        const transaction = parseTransactionInput(event, queueItem.transaction)
+        q.set(event.traceID, {
+          traceID: event.traceID,
+          workflow: 'TRANSACTION_REVIEW',
+          transaction,
+        })
+        dispatch({
+          type: 'UPDATE_TRANSACTION',
+          transaction,
+        })
+      }
+
+      break
+    }
+  }
+
+  return q
+}
+
+const getSplashClasses = (item?: QueueItem) => {
+  if (item && 'view' in item && item?.view === 'success') {
+    return {
+      background: 'bg-success',
+      color: 'text-black',
+    }
+  }
+  return {
+    background: 'bg-black',
+    color: 'text-white',
+  }
+}
+
+export const InteractionsProvider = () => {
+  const { service, dispatch } = useGlobal()
+  const [queue, setQueue] = useState<Queue>(new Map())
+  const currentFlowItem = queue.entries().next().value
+
+  const { background, color } = getSplashClasses(currentFlowItem?.[1])
+
+  const transitions = useTransition(!!currentFlowItem, {
+    from: { opacity: 0 },
+    enter: { opacity: 1 },
+    leave: { opacity: 0 },
+    config: { ...config.default, duration: 300 },
+  })
+
+  useEffect(() => {
+    service.EventsOn('new_interaction', (event: RawInteraction) => {
+      if (event.name === 'INTERACTION_SESSION_BEGAN') {
+        interactionBus.addListener(event.traceID, (event: RawInteraction) => {
+          setQueue((queue) => handleEvent({ queue, event, dispatch }))
+        })
+      }
+
+      interactionBus.emit(event.traceID, event)
+
+      if (event.name === 'INTERACTION_SESSION_ENDED') {
+        interactionBus.removeListener(event.traceID)
+      }
+    })
+
+    return () => service.EventsOff('new_interaction')
```

Comment:
I think we need to clear all event handlers here as well otherwise we will leak memory